packages/cli/src/rollup-plugins/server-components.ts

```diff
@@ -0,0 +1,75 @@
+import { AcornNode, Plugin, SourceDescription } from "rollup";
+import { Program } from "estree";
+import MagicString, { SourceMap } from "magic-string";
+
+export function serverComponentsPlugin({
+  sourceMap,
+}: {
+  sourceMap: boolean;
+}): Plugin {
+  return {
+    name: "server-components",
+
+    transform(code, id) {
+      if (code.includes("'use client'") || code.includes('"use client"')) {
+        const ast: Program = (() => {
+          const babelMeta = this.getModuleInfo(id)!.meta.babel;
+          if (babelMeta?.codeAtBabelTime === code) {
+            return this.getModuleInfo(id)!.meta.babel.ast;
+          }
+          return this.parse(code);
+        })();
+
+        for (const [idx, node] of ast.body.entries()) {
+          if (
+            node.type !== "ExpressionStatement" ||
+            node.expression.type !== "Literal" ||
+            typeof node.expression.value !== "string"
+          ) {
+            return;
+          }
+          if (node.expression.value === "use client") {
+            let magicString = new MagicString(code);
+            this.emitFile({ type: "chunk", id });
```

Comment:
I guess this is the core of the change that allows Rollup to preserve this file as a separate chunk, right?

Reply:
Yes